mosaic/mosaic.go

```diff
@@ -0,0 +1,423 @@
+	} else {
+		// Default to black if no foreground pixels.
+		block.BestFgColor = color.Black
+	}
+
+	if len(bgPixels) > 0 {
+		block.BestBgColor = m.averageColors(bgPixels)
+	} else {
+		// Default to black if no background pixels.
+		block.BestBgColor = color.Black
+	}
+
+	block.BestSymbol = bestChar
+}
+
+// averageColors calculates the average color from a slice of colors.
+func (m *Mosaic) averageColors(colors []color.Color) color.Color {
+	if len(colors) == 0 {
+		return color.Black
+	}
+
+	var sumR, sumG, sumB, sumA uint32
+
+	for _, c := range colors {
+		r, g, b, a := c.RGBA()
+		r, g, b, a = shift(r), shift(g), shift(b), shift(a)
+		sumR += r
+		sumG += g
+		sumB += b
+		sumA += a
+	}
+
+	count := uint32(len(colors)) //nolint:gosec
+	return color.RGBA{
+		R: uint8(sumR / count), //nolint:gosec
+		G: uint8(sumG / count), //nolint:gosec
+		B: uint8(sumB / count), //nolint:gosec
+		A: uint8(sumA / count), //nolint:gosec
+	}
+}
+
+// getPixelSafe returns the color at (x,y) or black if out of bounds.
+func (m *Mosaic) getPixelSafe(img image.Image, x, y int) color.RGBA {
+	bounds := img.Bounds()
+	if x < bounds.Min.X || x >= bounds.Max.X || y < bounds.Min.Y || y >= bounds.Max.Y {
+		return color.RGBA{0, 0, 0, 255}
+	}
+
+	r8, g8, b8, a8 := img.At(x, y).RGBA()
+	return color.RGBA{
+		R: uint8(r8 >> 8), //nolint:gosec
+		G: uint8(g8 >> 8), //nolint:gosec
+		B: uint8(b8 >> 8), //nolint:gosec
+		A: uint8(a8 >> 8), //nolint:gosec
+	}
+}
+
+// applyScaling resizes an image to the specified dimensions.
+func (m *Mosaic) applyScaling(img image.Image, width, height int) image.Image {
+	rect := image.Rect(0, 0, width, height)
+	dst := image.NewRGBA(rect)
+	xdraw.ApproxBiLinear.Scale(dst, rect, img, img.Bounds(), draw.Over, nil)
+	return dst
+}
+
+// applyDithering applies Floyd-Steinberg dithering.
+func (m *Mosaic) applyDithering(img image.Image) image.Image {
+	b := img.Bounds()
+	pm := image.NewPaletted(b, palette.Plan9)
+	draw.FloydSteinberg.Draw(pm, b, img, image.Point{})
+	return pm
+}
+
+// invertImage inverts the colors of an image.
+func (m *Mosaic) invertImage(img image.Image) image.Image {
+	bounds := img.Bounds()
+	width := bounds.Max.X - bounds.Min.X
+	height := bounds.Max.Y - bounds.Min.Y
+
+	result := image.NewRGBA(bounds)
+	for y := 0; y < height; y++ {
+		for x := 0; x < width; x++ {
+			r8, g8, b8, a8 := img.At(x+bounds.Min.X, y+bounds.Min.Y).RGBA()
+			result.Set(x, y, color.RGBA{
+				R: uint8(255 - (r8 >> 8)), //nolint:gosec
+				G: uint8(255 - (g8 >> 8)), //nolint:gosec
+				B: uint8(255 - (b8 >> 8)), //nolint:gosec
+				A: uint8(a8 >> 8),         //nolint:gosec
+			})
+		}
+	}
+
+	return result
+}
+
+// rgbaToLuminance converts RGBA color to luminance (brightness).
+func rgbaToLuminance(c color.Color) uint8 {
+	r, g, b, _ := c.RGBA()
+	r, g, b = shift(r), shift(g), shift(b)
+	// Weighted RGB to account for human perception.
+	return uint8(float64(r)*0.299 + float64(g)*0.587 + float64(b)*0.114)
```

Comment:
what are these constants? 

maybe make `const` for each each?

Reply:
Perhaps we can use `go-colorful` here

```go
co, _ := colorful.MakeColor(c)
// get HSL value
co.HSL()
```

Reply:
it's from https://www.w3.org/TR/AERT/#color-contrast, will add a comment about it